### PR TITLE
[Example] CompGCN

### PR DESCRIPTION
**Pull request by KounianhuaDu**
## Description
This is the implementation of compGCN. Compared to the MVP example, the main differences are:
- Single relation ->  Multi-relation
- Add in the basis vector
- Add in the in/out edge norm
- Dropouts in and after the compGCN layers (Following the paper, in the layer, the droput only acts on the in/out features but dose not act on the loop feature)
- Batchnorm (Following the paper, the batchnorm only acts on the node features but dose not act on the relation features)
- Activation: tanh (This only acts on node features too)
- node classification -> link prediction
- ConvE link prediction module

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by KounianhuaDu** [comment]
@zhjwy9343

File: examples/pytorch/compGCN/models.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        self.bn1 = th.nn.BatchNorm2d(self.num_filt)
+        self.bn2 = th.nn.BatchNorm1d(self.embed_dim)
+
+        self.hidden_drop = th.nn.Dropout(self.hid_drop)
+        self.feature_drop = th.nn.Dropout(self.feat_drop)
+        self.m_conv1 = th.nn.Conv2d(1, out_channels=self.num_filt, kernel_size=(self.ker_sz, self.ker_sz), stride=1, padding=0, bias=False)
+        
+        flat_sz_h = int(2*self.k_w) - self.ker_sz + 1
+        flat_sz_w = self.k_h - self.ker_sz + 1
+        self.flat_sz = flat_sz_h*flat_sz_w*self.num_filt
```
Comment (zhjwy9343) [inline examples/pytorch/compGCN/models.py:227]: add spaces between operators

File: examples/pytorch/compGCN/models.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        self.compGCN_Model = CompGCN(B, R, num_ent, in_dim, layer_size, comp_fn, batchnorm, dropout, layer_dropout)
+        
+        self.bn0 = th.nn.BatchNorm2d(1)
+        self.bn1 = th.nn.BatchNorm2d(self.num_filt)
+        self.bn2 = th.nn.BatchNorm1d(self.embed_dim)
+
+        self.hidden_drop = th.nn.Dropout(self.hid_drop)
+        self.feature_drop = th.nn.Dropout(self.feat_drop)
+        self.m_conv1 = th.nn.Conv2d(1, out_channels=self.num_filt, kernel_size=(self.ker_sz, self.ker_sz), stride=1, padding=0, bias=False)
+
```
Comment (zhjwy9343) [inline examples/pytorch/compGCN/models.py:224]: pls add some comments here to explain what is the purpose of these 4 lines.

File: examples/pytorch/compGCN/models.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        flat_sz_h = int(2*self.k_w) - self.ker_sz + 1
+        flat_sz_w = self.k_h - self.ker_sz + 1
+        self.flat_sz = flat_sz_h*flat_sz_w*self.num_filt
+        self.fc = th.nn.Linear(self.flat_sz, self.embed_dim) 
+
+        self.bias = nn.Parameter(th.zeros(num_ent))
+        
+    
+    def concat(self, e1_embed, rel_embed):
+        e1_embed = e1_embed. view(-1, 1, self.embed_dim)
```
Comment (zhjwy9343) [inline examples/pytorch/compGCN/models.py:234]: e1_embed. view -> e1_embed.view

**Review by zhjwy9343 — COMMENTED** [review]
In general, pls add more comments to help readers to understand the code logic.

File: examples/pytorch/compGCN/models.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+                comp_h_s = n_in_feats - r_feats[-1]
+            elif self.comp_fn == 'mul':
+                comp_h_s = n_in_feats * r_feats[-1]
+            elif self.comp_fn == 'ccorr':
+                comp_h_s = ccorr(n_in_feats, r_feats[-1])
+            else:
+                raise Exception('Only supports sub, mul, and ccorr')
+
+            # Sum all of the comp results as output of nodes and dropout
+            n_out_feats = (self.W_S(comp_h_s) + self.dropout(g.ndata['comp_edge'])) * (1/3)
```
[the remark this answers is not in the document]
Comment (KounianhuaDu) [reply examples/pytorch/compGCN/models.py:99]: It is 1/3 for both. It's like:
1/3*(a+b+c) = 1/3*a + 1/3*(b+c).
Here comp_fn = in + out        ---> b+c
         self-loop                        ---> a

**Comment by KounianhuaDu** [comment]
@zhjwy9343 The revision is done.

**Review by zhjwy9343 — APPROVED** [review]
LGTM
